### PR TITLE
Rewrite Flux Point energy bin computation

### PR DESCRIPTION
**Pull request by mealworm**
2018-07-10 coding sprint: fix bugs in flux point energy bin computation by simplifying code

### CODE REVIEW

File: gammapy/spectrum/energy_group_new.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    def upper_bounds(self):
+        return self.indices[1:] - 1
+
+    def bin(self, idx):
+        return np.arange(self.lower_bounds[idx],
+                         self.upper_bounds[idx])
+
+
+
+class FluxPointBinMaker:
```
Comment (mackaiver) [inline gammapy/spectrum/energy_group_new.py]: I'm a bit perplexed by the ubiquity of these `maker` classes within gammapy. Why doesn't this functionality simply go into the  FluxPointBins class?

This directly implies the next obvious question of why there is a class called `FluxPoints`but no class called `FluxPoint` which just knows its bin edges?

File: gammapy/spectrum/energy_group_new.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    def __init__(self, obs):
+        self.obs = obs
+        self.indices = None
+
+    def compute_bins_fixed(self, energy_binning):
```
Comment (mackaiver) [inline gammapy/spectrum/energy_group_new.py]: I think this class could be replaced by a single method.

File: gammapy/spectrum/energy_group_new.py (hunk 1 above)
Comment (mealworm) [reply gammapy/spectrum/energy_group_new.py]: I completely agree about Maker classes. To me it makes python look like Java. I would have just made an initialization method to FluxPointBins, but was told that separate maker classes are the house style in gammapy. Is that not the case?

Comment (mealworm) [reply gammapy/spectrum/energy_group_new.py]: Regarding FluxPoint, that adds some complexity. Really this whole class is just a wrapper around a single array of indices. FluxPoint would then be a wrapper around just one index.

File: gammapy/spectrum/energy_group_new.py (hunk 2 above)
Comment (mealworm) [reply gammapy/spectrum/energy_group_new.py]: The rationale I was told was that in the future there will be other ways of defining bins without first going through the energy values. But the same thing can be accomplished just by overloading the init method, so I agree.

File: gammapy/spectrum/energy_group_new.py (hunk 1 above)
Comment (mackaiver) [reply gammapy/spectrum/energy_group_new.py]: > Regarding FluxPoint, that adds some complexity. Really this whole class is just a wrapper around a single array of indices. FluxPoint would then be a wrapper around just one index.

I agree. It was a somewhat suggestive question. It seems to me like its a bit of OO overkill in this case.

> To me it makes python look like Java.
Exactly. That's what I was reminded of  :smile:

Comment (mackaiver) [reply gammapy/spectrum/energy_group_new.py]: > but was told that separate maker classes are the house style in gammapy. Is that not the case?

I don't know to be honest. Maybe @cdeil can chime in.

Comment (mealworm) [reply gammapy/spectrum/energy_group_new.py]: Sorry, one pair of indices (obviously)

**Comment by cdeil** [comment]
@mealworm - Thanks for the PR and @mackaiver for the review!

@joleroi - Assigning to you for final review / merging.

I would suggest that we discuss this a bit tomorrow morning before we decide what to put.

One point in mind that the main use case for this is to support computation of spectral points as shown [here](http://docs.gammapy.org/dev/notebooks/cta_data_analysis.html#Spectral-points) using the [SpectrumEnergyGroupMaker](http://docs.gammapy.org/dev/api/gammapy.spectrum.SpectrumEnergyGroupMaker.html).
One option is to keep the API of that class to keep existing scripts running, just replace the implementation with this one and delete the same code.
We can also put a new API if it's clear that the new one will be better, but this needs some thought / discussions (as already started by @mackaiver and @mealworm above).

**Comment by mealworm** [comment]
After discussion, it was decided to fix the existing classes rather than write a new one.

**Pull request closed without merging**